tests/test_device_attach_notification_nordy/src/test_device_attach_notification_nordy.xc

```diff
@@ -0,0 +1,180 @@
+// Copyright 2016-2025 XMOS LIMITED.
+// This Software is subject to the terms of the XMOS Public Licence: Version 1.
+
+#include <xs1.h>
+#include <print.h>
+#include <stdio.h>
+#include "xud.h"
+#include "platform.h"
+#include "xud_shared.h"
+
+#define XUD_EP_COUNT_OUT    2
+#define XUD_EP_COUNT_IN     2
+
+#ifndef TEST_EP_NUM
+#define TEST_EP_NUM         1
+#endif
+
+#define PKT_LENGTH          10
+
+#ifndef XUD_TEST_SPEED
+#error XUD_TEST_SPEED not defined
+#endif
+
+/* Endpoint type tables */
+XUD_EpType epTypeTableOut[XUD_EP_COUNT_OUT] = {XUD_EPTYPE_CTL, XUD_EPTYPE_BUL | XUD_STATUS_ENABLE};
+XUD_EpType epTypeTableIn[XUD_EP_COUNT_IN] =   {XUD_EPTYPE_CTL, XUD_EPTYPE_BUL | XUD_STATUS_ENABLE};
+
+#pragma unsafe arrays
+testResult_t TestEpBusUpdate_Tx(XUD_ep ep_out0, XUD_ep ep_in, int epNum, int pktLen, t_runMode runMode)
+{
+    XUD_Result_t result;
+    XUD_BusSpeed_t busSpeed;
+    unsigned length;
+
+    unsigned char bufferRx[1024];
+    unsigned char bufferTx[1024];
+    unsigned char bufferTxBad[1024];
+
+    set_core_fast_mode_on();
+
+    /* Prepare packet */
+    for(int i = 0; i < pktLen; i++)
+    {
+        bufferTx[i] = i;
+        bufferTxBad[i] = 0xBA;
+    }
+
+    /* At this point (due to test packet sequencing and code ordering) that XUD has already sent
+     * a bus reset notification. This packet should never make it to the host */
+    result = XUD_SetBuffer(ep_in, bufferTxBad, pktLen);
+
+    /* Since device has received a reset, don't expect the old buffer we set abouve to actually
+     * be transmitted. The EP should now be NAKing */
+
+    /* Sync with the host via EP0 so we know we can now get the bus status update and set some fresh
+     * data */
+    result = XUD_GetBuffer(ep_out0, bufferRx, length);
+    busSpeed = XUD_ResetEndpoint(ep_in, null);
+
+    if(busSpeed != XUD_TEST_SPEED)
+        return FAIL_BAD_BUS_SPEED;
+
+    /* Mark EP ready with a fresh new buffer */
+    result = XUD_SetBuffer(ep_in, bufferTx, pktLen);
+
+    /* Allow a little time for Tx data to make it's way of the port - important for FS tests */
+    {
+        timer t;
+        unsigned time;
+        t :> time;
+        t when timerafter(time + 500) :> int _;
+    }
+
+    if(runMode != RUNMODE_DIE)
+        while(1);
+
+    return 0;
+}
+
+testResult_t TestEpBusUpdate_Rx(XUD_ep ep_out0, XUD_ep ep_out1, int epNum, int expectedLength)
+{
+    XUD_Result_t result;
+    XUD_BusSpeed_t busSpeed = XUD_TEST_SPEED;
+
+    /* Buffer for packet data and lengths */
+    unsigned char buffer[1024];
+    unsigned length;
+
+    set_core_fast_mode_on();
+
+    /* Get packet on EP 0 such that we know host has end reset */
+    result = XUD_GetBuffer(ep_out0, buffer, length);
```

Comment:
I dont think I understand this. How are we getting away with calling XUD_GetBuffer(ep_out0, ... from both TestEpBusUpdate_Rx and TestEpBusUpdate_Tx? I feel like you've done some clever synchronisation that I'm not getting. I guess TestEpBusUpdate_Rx is receiving the 1st EP0 OUT packet that the host sends and TestEpBusUpdate_Tx is getting the 2nd EP0 OUT packet, but how?

Reply:
These functions are not run in parallel, but one after another. Does that help?

Reply:
See L150

Reply:
that makes sense! thanks.